go/consensus/tendermint/apps/beacon/backend_insecure.go

```diff
@@ -0,0 +1,228 @@
+	ctx *api.Context,
+	state *beaconState.MutableState,
+	params *beaconState.ConsensusParameters,
+	req types.RequestBeginBlock,
+) error {
+	future, err := state.GetFutureEpoch(ctx)
+	if err != nil {
+		return fmt.Errorf("beacon: failed to get future epoch: %w", err)
+	}
+	if future == nil {
+		return nil
+	}
+
+	height := ctx.BlockHeight() + 1 // Current height is ctx.BlockHeight() + 1
+	switch {
+	case future.Height < height:
+		// What the fuck, we missed transitioning the epoch?
+		ctx.Logger().Error("height mismatch in defered set",
+			"height", height,
+			"expected_height", future.Height,
+		)
+		return fmt.Errorf("beacon: height mismatch in defered set")
+	case future.Height > height:
+		// The epoch transition is scheduled to happen in the grim
+		// darkness of the far future.
+		return nil
+	case future.Height == height:
+		// Time to fire the scheduled epoch transition.
+	}
+
+	// Transition the epoch.
+	ctx.Logger().Info("setting epoch",
+		"epoch", future.Epoch,
+		"current_height", height,
+	)
+
+	if err = state.SetEpoch(ctx, future.Epoch, height); err != nil {
+		return fmt.Errorf("beacon: failed to set epoch: %w", err)
+	}
+	if err = state.ClearFutureEpoch(ctx); err != nil {
+		return fmt.Errorf("beacon: failed to clear future epoch: %w", err)
+	}
+	if !impl.isMockEpochTime(params) {
+		if err = impl.scheduleEpochTransitionBlock(ctx, state, params.Parameters.InsecureParameters, future.Epoch+1); err != nil {
+			return err
+		}
+	}
+	impl.app.doEmitEpochEvent(ctx, future.Epoch)
+
+	// Generate the beacon
+	return impl.onEpochChangeBeacon(ctx, state, params.Parameters.InsecureParameters, future.Epoch, req)
+}
+
+func (impl *backendInsecure) scheduleEpochTransitionBlock(
+	ctx *api.Context,
+	state *beaconState.MutableState,
+	params *beacon.InsecureParameters,
+	nextEpoch beacon.EpochTime,
+) error {
+	// Schedule the epoch transition based on block height.
+	nextHeight := int64(nextEpoch) * params.Interval
+	return impl.app.scheduleEpochTransitionBlock(ctx, state, nextEpoch, nextHeight)
+}
+
+func (impl *backendInsecure) onEpochChangeBeacon(
+	ctx *api.Context,
+	state *beaconState.MutableState,
+	params *beacon.InsecureParameters,
+	epoch beacon.EpochTime,
+	req types.RequestBeginBlock,
+) error {
+	var entropyCtx, entropy []byte
+
+	switch params.DebugDeterministic {
+	case false:
+		entropyCtx = prodEntropyCtx
+
+		height := ctx.BlockHeight()
+		if height <= ctx.InitialHeight() {
+			// No meaningful previous commit, use the block hash.  This isn't
+			// fantastic, but it's only for one epoch.
+			ctx.Logger().Debug("onBeaconEpochChange: using block hash as entropy")
+			entropy = req.Hash
+		} else {
+			// Use the previous commit hash as the entropy input, under the theory
+			// that the merkle root of all the commits that went into the last
+			// block is harder for any single validator to game than the block
+			// hash.
+			//
+			// TODO: This still isn't ideal, and an entirely different beacon
```

Comment:
This comment should probably be removed as this PR adds a SCRAPE-based beacon :-)